src/GroupDetails.ts

```diff
@@ -0,0 +1,205 @@
+            this.groups.set([], tasks);
+        } else {
+            const groupers = Group.getGroupersForGroups(grouping);
+
+            for (const task of tasks as Task[]) {
+                const groupNames = Group.getGroupNamesForTask(groupers, task);
+                this.addTask(groupNames, task);
+            }
+            this.groups = this.getSortedGroups();
+        }
+    }
+
+    private addTask(groupNames: string[], task: Task) {
+        const groupForNames = this.getOrCreateGroupForGroupNames(groupNames);
+        groupForNames?.push(task);
+    }
+
+    /**
+     * If a task has been seen before with this exact combination of group names,
+     * return the container that the previous task(s) were added to.
+     *
+     * Otherwise, create and return a new container.
+     * @param newGroupNames
+     * @private
+     */
+    private getOrCreateGroupForGroupNames(newGroupNames: string[]) {
+        for (const [groupNames, taskGroup] of this.groups) {
+            // Is there a better way to check if the contents of two string arrays
+            // are identical?
+            // Use of JSON feels inefficient, and is O(n-squared) on the number
+            // of unique group-name combinations, so may scale badly if
+            // very large numbers of tasks are displayed.
+            // Related: https://stackoverflow.com/questions/7837456/how-to-compare-arrays-in-javascript
+            if (JSON.stringify(groupNames) === JSON.stringify(newGroupNames)) {
+                return taskGroup;
+            }
+        }
+        const taskGroup: Task[] = [];
+        this.groups.set(newGroupNames, taskGroup);
+        return taskGroup;
+    }
+
+    private getSortedGroups() {
+        // groups.keys() will initially be in the order the entries were added,
+        // so effectively random.
+        // Return a duplicate map, with the keys (that is, group names) sorted in alphabetical order:
+        return new IntermediateTaskGroupsStorage(
+            [...this.groups.entries()].sort(),
+        );
+    }
+}
+
+/**
+ * GroupHeadings calculates which headings need to be displayed, for
+ * a given group of tasks.
+ *
+ * See the explanation in GroupDetails.ts for how it works.
+ */
+export class GroupHeadings {
+    private lastHeadingAtLevel = new Array<string>();
+
+    constructor(groupedTasks: IntermediateTaskGroupsStorage) {
+        const firstGroup = groupedTasks.keys().next().value;
+        const groupCount = firstGroup.length;
+        for (let i = 0; i < groupCount; i++) {
+            this.lastHeadingAtLevel.push('');
+        }
+    }
+
+    /**
+     * Calculate the minimal set of headings that should be displayed
+     * before the tasks with the given group names.
+     *
+     * Data for each required heading is stored in a GroupHeading object.
+     * @param groupNames 0 or more group names, one per 'group by' line
+     */
+    getHeadingsForTaskGroup(groupNames: string[]): GroupHeading[] {
```

Comment:
Are you calling this for subsequent groups while you retain state in `lastHeadingAtLevel` in between calls to `getHeadingsForTaskGroup`? That's a bad idea, as you don't know whether the state of `lastHeadingAtLevel` was in the meantime changed by something else, e.g. another call to `getHeadingsForTaskGroup` from somewhere else.

Keeping and relying on state like this produces very hard to find bugs. It also makes the code hard to read. In this method, you only iterate over `groupNames` once, so every `level` will exist only once. Yet you make a comparison, which took me a long time to figure out how that works.

Instead, it should be written in a way where the state must not be kept across multiple function calls. For example by iterating over all groups once and setting the headers for all groups in that one call.

I also _think_ it might be easier to figure out the headings? Though I am not sure.

Reply:
@schemar This is a key comment. Your understanding is correct, and doing what you suggest would simplify the code considerably, probably allowing the `GroupHeadings` and `IntermediateTaskGroups` classes to be merged in to the main `Group` class.

Just to explain why I did it this way. Imagine:

-  a vault with many, many thousands of notes, 
- and many, many thousands of tasks ... 
- and a query with several `group by` commands

I was concerned about the impact of QueryRender waiting for a big container to be created, containing multiple copies of the strings for group names...

I figured that it was lower-risk, performance-wise, to only create the list of headings to display iteratively.

Reply:
> I figured that it was lower-risk, performance-wise, to only create the list of headings to display iteratively.

Resolving conversation, as per comments above. Also, the code is likely to be restructured following other feedback anyway, so this may all go away,